File: 198-house-robber/198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        @cache
        def helper(pos: int) -> int:
            if pos < 0:
                return 0
            
            return max(
                nums[pos] + helper(pos - 2),
                helper(pos - 1)
            )
        n = len(nums)
        return helper(n - 1)
        

File: 198-house-robber/198-house-robber.java
class Solution {
    public int rob(int[] nums) {
        int prev = 0, curr = 0;
        for(int i = nums.length - 1; i >= 0; i--){
            int tmp = prev;
            prev = curr;
            curr = Math.max(nums[i] + tmp, curr);
        }
        return curr;
    }
}